panther_driver/src/panther_can.py

```diff
@@ -0,0 +1,195 @@
+#!/usr/bin/python3
+
+import canopen
+from time import time, sleep
+from threading import Thread
+
+import rospy
+
+from constants import LEFT_WHEEL, RIGHT_WHEEL, VOLT_CHANNEL, AMP_CHANNEL
+
+
+class MotorController:
+    def __init__(self, can_node_id, eds_file) -> None:
+        self.wheel_pos = [0.0, 0.0]
+        self.wheel_vel = [0.0, 0.0]
+        self.wheel_curr = [0.0, 0.0]
+        self.battery_data = [0.0, 0.0] # V, I
+        self.runtime_stat_flag = [0, 0]
+        self.fault_flags = 0
+
+        self.can_node = canopen.RemoteNode(can_node_id, eds_file)
+
+
+class PantherCAN:
+    def __init__(self, eds_file, can_interface) -> None:
+        self.lock = False
+        self.can_net_err = False
+        self._max_err_per_sec = 2
+        self._err_times = [0] * self._max_err_per_sec
+        
+        self._network = canopen.Network()
+
+        self._motor_controllers = [
+            MotorController(1, eds_file),   # front
+            MotorController(2, eds_file)    # rear
+        ]
+        
+        self._network.connect(channel=can_interface, bustype='socketcan')
+
+        self._network.add_node(self._motor_controllers[0].can_node)
+        self._network.add_node(self._motor_controllers[1].can_node)
+        
+        self._connection_check_timer = Thread(target=self._can_net_check)  
+        self._connection_check_timer.start()
+
+        rospy.loginfo(f'[{rospy.get_name()}] Connected to the CAN bus.')
+        
+    def set_wheels_enc_velocity(self, vel: list) -> None:
+        while self.lock: 
+            sleep(0.01)
+            
+        self.lock = True
+        for motor_controller, enc_vel in zip(self._motor_controllers, [vel[:2], vel[2:]]):
+            for i, wheel in enumerate([LEFT_WHEEL, RIGHT_WHEEL]):
+                try:
+                    motor_controller.can_node.sdo['Cmd_CANGO'][wheel].raw = enc_vel[i]
+                except:
+                    rospy.logwarn(
+                        f'[{rospy.get_name()}] PantherCAN: SdoCommunicationError '
+                        f'occurred while setting wheels velocity'
+                    )
+                    self._error_handle()
+        self.lock = False
```

Comment:
Using normal lock this would look like this:
```suggestion
    def set_wheels_enc_velocity(self, vel: list) -> None:
        with self.lock:
          for motor_controller, enc_vel in zip(self._motor_controllers, [vel[:2], vel[2:]]):
              for i, wheel in enumerate([LEFT_WHEEL, RIGHT_WHEEL]):
                  try:
                      motor_controller.can_node.sdo['Cmd_CANGO'][wheel].raw = enc_vel[i]
                  except:
                      rospy.logwarn(
                          f'[{rospy.get_name()}] PantherCAN: SdoCommunicationError '
                          f'occurred while setting wheels velocity'
                      )
                      self._error_handle()
```
```